README.md
Homebrew IWYU
=============

[![Build Status](https://img.shields.io/travis/jasonmp85/homebrew-iwyu/master.svg)][status]
[![Release](https://img.shields.io/github/release/jasonmp85/homebrew-iwyu.svg)][release]
[![License](https://img.shields.io/:license-mit-blue.svg)][license]

This formula makes it easy to install `include-what-you-use` on any modern OS X system.

Just `brew tap jasonmp85/iwyu` and then `brew install iwyu`.

Using `iwyu`
------------

The [project's page][iwyu-page] goes into more detail, but there are three basic ways to use `iwyu`…

### Directly

Invoke it on a single file, as you would a compiler: `iwyu hello_world.c`. Messages about what includes to add or remove will be printed to standard output.

### From `make`

Tell make to use it as the C compiler: `make -k CC=iwyu`. It's necessary to use the `-k` flag to continue after errors (`iwyu` always errors to signal that no compilation has actually taken place).

### Using `fix_include`

`include-what-you-use` bundles a Python script capable of parsing its output in order to automatically fix any include problems, in-place if you desire. It's not perfect, but something like `fix_include hello_world.c < iwyu hello_world.c` should work to update a file named _hello_world.c_ with the suggestions made by `iwyu`.

Copyright
---------

Copyright © 2014–2017 Jason Petersen

Code released under the [MIT License](LICENSE).

[status]: https://travis-ci.org/jasonmp85/homebrew-iwyu
[release]: https://github.com/jasonmp85/homebrew-iwyu/releases/latest
[license]: LICENSE
[iwyu-page]: http://include-what-you-use.org


CHANGELOG.md
### homebrew-iwyu v5.0.0 (January 11, 2017) ###

* Installs include-what-you-use 0.7 (based on clang 3.8.0)

* Removes brittle Xcode-symlink workaround

* Relies on compiler builtin headers included in iwyu

* Relies on iwyu's new macOS header path logic

* Adds Xcode 8.0, 8.1, and 8.2 to build matrix

* Removes pre-7.3 Xcode versions from build matrix

### homebrew-iwyu v4.0.0 (June 16, 2016) ###

* Installs include-what-you-use 0.6 (based on clang 3.8.0)

* Compiles against Apple LLVM 7.3.0

* Adds Xcode 7.3 to Travis build matrix

### homebrew-iwyu v3.0.0 (December 23, 2015) ###

* Installs include-what-you-use 0.5 (based on clang 3.7.0)

* Drops support for Xcode 6; now requires 7.0 or higher

* Removes default OS X (Xcode 6.1) build from Travis

### homebrew-iwyu v2.1.0 (October 15, 2015) ###

* Supports Xcode 6.1 or higher, including 7.0

* Determines Xcode version at build time

* Adds support for C++ builtin headers

* No longer pollutes system include dirs

* Builds against multiple Xcode versions in Travis

* Adds stronger unit tests (fewer false positives)

### homebrew-iwyu v2.0.0 (June 28, 2015) ###

* Installs include-what-you-use 0.4

* Requires Xcode 6.1 or higher

* Compiles against LLVM 3.6

* Updates project page links

* Improves Travis CI build

### homebrew-iwyu v1.0.0 (December 6, 2014) ###

* Uses prebuilt tarball instead of compiling from source

* Symlinks to Xcode's clang headers rather than bringing in LLVM

* Requires Xcode 6.0 or higher

### homebrew-iwyu v0.9.0 (November 19, 2014) ###

* Initial release

* Builds `clang_3.5` tag from Subversion

* Compiles against LLVM 3.5


iwyu.rb
require "English"
require "formula"

# include-what-you-use needs access to headers included with LLVM 3.8, which
# is only present in Xcode version 7.0 or higher.
class Xcode70 < Requirement
  fatal true

  satisfy { MacOS::Xcode.version >= "7.0" }

  def message
    "Xcode 7.0 or newer is required for this package."
  end
end

# This formula provides an easy way to install include-what-you-use, a tool for
# automatically managing include directives within C and C++ projects. It will
# build and install include-what-you-use, symlink it as iwyu, and install a
# Python wrapper to automatically correct includes (fix_include).
class Iwyu < Formula
  # iwyu 0.6 based on clang 3.8
  CLANG_VERSION = "3.8".freeze

  version "0.6"
  homepage "http://include-what-you-use.org"
  url "http://include-what-you-use.org/downloads/" \
      "include-what-you-use-#{version}-x86_64-apple-darwin.tar.gz"
  sha256 "46a7e579ad17441ba8b23fe105400565ff978cc61957c19291563912d1f4638d"

  depends_on Xcode70

  def install
    # include-what-you-use needs lib and include directories one level
    # up from its bindir, but putting them in the homebrew directories
    # results in them getting linked in /usr (interfering with e.g. the
    # /usr/local/include dirs added by gcc).
    #
    # To solve this, we put iwyu in a custom dir and make bin links.
    iwyu_subdir_path = (prefix / "iwyu")
    iwyu_subdir_path.mkpath

    clang_version = `clang --version`.each_line.first
    apple_llvm_version = clang_version[/\AApple LLVM version (\d+(?:\.\d+)+)/, 1]

    clang_libs = "#{MacOS::Xcode.toolchain_path}/usr/lib/clang/#{apple_llvm_version}"
    cpp_includes = "#{MacOS::Xcode.toolchain_path}/usr/include/c++"

    iwyu_bindir = (iwyu_subdir_path / "bin")
    iwyu_libdir = (iwyu_subdir_path / "lib")
    iwyu_includes = (iwyu_subdir_path / "include")

    iwyu_bindir.mkpath
    iwyu_libdir.mkpath
    iwyu_includes.mkpath

    iwyu_clang_lib_path = (iwyu_libdir / "clang")
    iwyu_clang_lib_path.mkpath
    iwyu_clang_lib_path.install_symlink(clang_libs => "#{Iwyu::CLANG_VERSION}.0")

    iwyu_includes.install_symlink(cpp_includes => "c++")

    iwyu_bindir.install("bin/fix_includes.py" => "fix_include")
    iwyu_bindir.install("bin/include-what-you-use")
    iwyu_bindir.install_symlink("include-what-you-use" => "iwyu")

    bin.install_symlink Dir["#{iwyu_bindir}/*"]
  end

  test do
    # write out a header and a C++ file relying on transitive dependencies
    (testpath / "demo.hpp").write <<-EOS.undent
    #include <stdio.h>
    #include <stdarg.h>
    #include <locale>
    EOS

    (testpath / "demo.cpp").write <<-EOS.undent
    #include "demo.hpp"

    int main(void)
    { printf("hello world"); }
    EOS

    # iwyu exits with a code equal to the number of suggested edits + 2
    fixes = shell_output "iwyu #{testpath}/demo.cpp 2>&1", 6
    assert_not_match(/file not found/, fixes)

    # pass the output to the fixer script and assert that it fixed two files
    results = pipe_output "fix_include", fixes
    assert_match(/IWYU edited 2 files/, results)

    # sigh. they use the status code to signal how many files were edited
    assert_equal 2, $CHILD_STATUS.exitstatus
  end

  def caveats; <<-EOS.undent
    This package will break after an Xcode upgrade. Fixing it is as simple as:
      brew reinstall iwyu
    EOS
  end
end
